### ISSUE TITLE
What is the purpose of hotfix tags in master branch?

### ISSUE THREAD

**Issue by thomasgi1** [petervanderdoes/gitflow-avh#365]
IMHO tagging hotfixes on the master branch makes no sense. I thought the purpose of these tags is to be able to checkout the sources of any released software later on (e.g. when investigating bug reports). Especially when you create a hotfix, the released software is based on the last commit in the hotfix branch. When you finish the hotfix, the changes of the hotfix are merged back into the master and develop branches and the master branch is tagged with the hotfix version. But the merge commit of the master branch may differ from the last commit of the hotfix branch! The latter is the case, if the hotfix is not based on the latest release, but on one of the previous releases. This is common scenario, if you have to maintain legacy releases. Furthermore, if you use git flow AVH with the default settings, then the hotfix branch will be deleted even on the remote repository. So you lose any track of the sources used to build the hotfix. The tag on the master branch is useless, because the master branch already contains changes included in the latest release.
I know that some previous versions of git flow already tagged the release/hotfix branches instead of the master branch, but there were some arguments of the community against this change and obviously it was reverted, which I did not understand.
So, how the current hotfix tagging is supposed to work? And if it does not work in my described scenario, how about an additional git flow option to choose the location of the tags?

**Comment by thomasgi1** [comment]
Thanks for clarification. Since the support branch was a experimental feature I never looked at its capabilities. I will evaluate that.
But one issue with the tags remains: We use git flow in conjunction of Jenkins Multibranch Pipelines. When we prepare a release or hotfix, then Jenkins automatically builds the binaries of the branch and triggers all the automatic tests. Also the manual test are done with binaries created out of the release/hotfix branch. Once all system tests succeeded we declare the latest build of the release/hotfix branch as official release, i.e. we do not finish the release/hotfix first and then build the master branch in order to get the final binaries. If we would that, then we would need test the master branch binaries again, but for the master branch we have only a single shot (the merge commit). If that fails for some reason (e.g. a Makefile problem that only occurs in master branch because of the different branch name pattern that generates version information), then we would need to start a complete new release cycle. Thus our current workflow is to keep the release/hotfix branches, so that we can recreate the binaries later based on the branch name. A tag on the last commit of the release/hotfix branch would be better for that purpose.

**Comment by jkankiewicz** [comment]
> Once all system tests succeeded we declare the latest build of the release/hotfix branch as official release, i.e. we do not finish the release/hotfix first and then build the master branch in order to get the final binaries.

@thomasgi1, You're not really using Gitflow then, are you? After all, release and hotfix branches are by their very nature ephemeral and only the master, develop and support branches are meant to endure. A git repository should never be polluted with old release and hotfix branches when the Gitflow branching model is being used.
Why can't you make the Jenkins job responsible for finishing the release/hotfix branch based on the validity of the binaries that it built? That way, a branch will never be finished unless its corresponding binaries are valid and the job can use the resultant tag's name to identify the official release. This approach would still avoid redundant builds and validation testing per release and keep your git repository clean. You could use the [input step](https://jenkins.io/doc/pipeline/steps/pipeline-input-step) of the Jenkins pipeline to integrate your manual testing tasks into the success of the job.

**Comment by thomasgi1** [comment]
Of course we could instruct Jenkins to finish the release branch once all tests succeeded. In fact we are doing this manually now. But that is not the point. The problem is that we are testing binaries built from the release branch and not from the master branch, when the release was finished. Yes, in theory the last commit of the release branch and the merge commit into the master branch are identical. But a product that we ship to customers is a bit more than pure source files. That is we only trust the tested binaries. And these binaries were generated from the release branch. And we want to freeze the related sources, which generated the binaries (including the Jenkinsfile used in the release branch).
After thinking about you suggestion to keep the release branch open I got the idea to let Jenkins finish the release branch without pushing the results to the remote repository. Then the master branch is built and tested. If the tests fail, then both the master and release branch are reset to the previous commit, so that we have a chance to fix the issue before we can do a next try to finish the release. Once the test succeeded Jenkins could push the tagged master branch. This procedure sounds doable, but is more complicated and is far away from the git flow idea.

**Comment by jkankiewicz** [comment]
> The problem is that we are testing binaries built from the release branch and not from the master branch, when the release was finished.

That's an artificial problem; the only way to change the master branch is by merging a hotfix or release branch into it and that means that there's no qualitative difference between the source code in a hotfix or release branch and the source code in the master branch after that hotfix or release branch is merged into it.

> Yes, in theory the last commit of the release branch and the merge commit into the master branch are identical.

It's not a theory, it's a fact.

**Comment by jkankiewicz** [comment]
> Fact is that we already had a difference between the release branch and the master branch, which caused some trouble. The reason for the difference was a wrong push into the master branch by a developer. OK, as consequence we protected the master branch now in GitLab, so that only an admin can push into the master branch. But the admin is just a human... 

Jenkins isn't human. If the Jenkins job that builds the binaries was also responsible for merging the hotfix/release branch into master and it logged into git using an admin account then you could be certain that the version tag for the release corresponds exactly to the binaries and there'd be no lingering hotfix/release branches polluting your repository.
> However, as I mentioned before, the content of the binaries not only depend on the sources. For instance the branch name itself has influence on the binaries, because the compiled-in version information is derived from the branch name.

A release branch's name should match the pattern "release/_<version>_" and the tag on the commit of its merge into master should be "_<version>_" so if the compiled -in version number is "_<version>_" then the tag on the merge commit is no worse than the release branch's name.

**Comment by Finesse** [comment]
Because version tags are given only to releases and you release only the code located on the `master` branch (it is called `master` for this reason) that's why you can have version tags only on `master`. If you need a pointer to a hotfix branch, just leave the hotfix branch.
